python/cudf/cudf/core/dataframe.py

```diff
@@ -8079,6 +8079,95 @@ def value_counts(
+
+        Parameters
+        ----------
+        copy : bool
+            Whether or not to generate a new copy of the underlying device data
+
+        Returns
+        -------
+        pylibcudf.Table
+            A pylibcudf.Table referencing the same data.
+        dict
+            Dict of metadata (includes column names and dataframe indices)
+
+        Notes
+        -----
+        User requests to convert to pylibcudf must assume that the
+        data may be modified afterwards.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> df = cudf.DataFrame({'a': [1, 2, 3],
+        ...                    'b': ['a', 'b', 'c']},
+        ...                    index=['x', 'y', 'z'])
+        >>> df.to_pylibcudf()
+        (<pylibcudf.table.Table at 0x7f9b5996e1d0>,
+        {'index': Index(['x', 'y', 'z'], dtype='object'),
+        'column_names': Index(['a', 'b'], dtype='object')})
+        """
+        if copy:
+            raise NotImplementedError("copy=True is not supported")
+        metadata = {"index": self.index, "columns": self.columns}
+        return plc.Table(
+            [col.to_pylibcudf(mode="write") for col in self._columns]
+        ), metadata
+
+    @classmethod
+    @_performance_tracking
+    def from_pylibcudf(cls, table: plc.Table, metadata: dict) -> Self:
+        """
+        Create a DataFrame from a pylibcudf.Table.
+
+        Parameters
+        ----------
+        table : pylibcudf.Table
+            The input Table.
+        metadata : dict
+            Metadata necessary to reconstruct the dataframe
+
+        Returns
+        -------
+        table : cudf.DataFrame
+            A cudf.DataFrame referencing the columns in the pylibcudf.Table.
+        metadata : list[str]
+            Dict of metadata (includes column names and dataframe indices)
+
+        Notes
+        -----
+        This function will generate a DataFrame which contains a tuple of columns
+        pointing to the same columns the input table points to.  It will directly access
+        the data and mask buffers of the pylibcudf columns, so the newly created
+        object is not tied to the lifetime of the original pylibcudf.Table.
+        """
+        if (
+            not isinstance(metadata, dict)
+            or len(metadata) == 0
+            or len(metadata) > 2
+            or "columns" not in metadata
+            or (len(metadata) == 2 and {"columns", "index"} != set(metadata))
+        ):
+            raise ValueError(
+                "Must at least pass metadata dict with column names and optionally indices only"
+            )
+        columns = table.columns()
+        df = cls._from_data(
+            {
+                name: cudf.core.column.ColumnBase.from_pylibcudf(
+                    col, data_ptr_exposed=True
+                )
+                for name, col in zip(metadata["columns"], columns)
+            },
+            index=metadata["index"],
```

Comment:
From the error message above, will we allow `"index"` to be optional? If so, this should be `metadata.get("index")`

Reply:
Yep, thanks